### PR TITLE
Fix build on s390x; related to #166

### PR DESCRIPTION
**Pull request by ueno**
Even after #166 got merged, the NSS build is still failing on s390x as mentioned in:
https://bugzilla.mozilla.org/show_bug.cgi?id=1615557

The architecture has [several differences from other arches](https://fedoraproject.org/wiki/Architectures/s390x), but with this change, the build succeeds:
* https://src.fedoraproject.org/rpms/nss/blob/master/f/nss-kremlin-ppc64le.patch
* https://koji.fedoraproject.org/koji/buildinfo?buildID=1463542

### CODE REVIEW

File: include/kremlin/internal/types.h [hunk 1]
```diff
@@ -58,7 +58,8 @@ typedef const char *Prims_string;
 typedef __m128i FStar_UInt128_uint128;
 #elif !defined(KRML_VERIFIED_UINT128) && !defined(_MSC_VER) && \
       (defined(__x86_64__) || defined(__x86_64) || defined(__aarch64__) || \
-      (defined(__powerpc64__) && defined(__LITTLE_ENDIAN__)))
```
Comment (franziskuskiefer) [inline include/kremlin/internal/types.h]: With this getting more and more complex we should maybe add a custom define for this so we have only one place where this is specified.

**Comment by beurdouche** [comment]
I suggest we do, I suspect these macros are breaking the Firefox build that I am debugging right now, so regrouping them would be useful...

**Comment by beurdouche** [comment]
I think I like that.

**Comment by ueno** [comment]
Thank you for the suggestion; although I haven't tested it yet, it might be something like this?

File: include/kremlin/internal/types.h [hunk 2]
```diff
@@ -75,15 +77,13 @@ typedef FStar_UInt128_uint128 FStar_UInt128_t, uint128_t;
...
+#if !defined(KRML_VERIFIED_UINT128) && HAS_INT128
 #include "fstar_uint128_gcc64.h"
-#endif
-
-#if !defined(KRML_VERIFIED_UINT128) && defined(_MSC_VER) && !defined(__clang__)
+#if !defined(KRML_VERIFIED_UINT128) && IS_MSVC64
 #include "fstar_uint128_msvc.h"
 #elif defined(KRML_VERIFIED_UINT128)
 #include "FStar_UInt128_Verified.h"
+#include "fster_uint128_struct_endianness.h"
```
Comment (franziskuskiefer) [inline include/kremlin/internal/types.h]: typo `fstar_uint128_struct_endianness`

File: include/kremlin/internal/types.h [hunk 3]
```diff
@@ -49,19 +49,21 @@ typedef const char *Prims_string;
 
 /* The great static header headache. */
 
+#define IS_MSVC64 (defined(_MSC_VER) && defined(_M_X64) && !defined(__clang__))
+#define HAS_INT128 ((defined(__x86_64__) || defined(__x86_64) || \
+		     defined(__aarch64__) || \
```
Comment (franziskuskiefer) [inline include/kremlin/internal/types.h]: Can you adjust indentation here to make this a little easier to read (the `defined` is in that bracket above).

File: include/kremlin/internal/types.h (hunk 2 above)
Comment (ueno) [reply include/kremlin/internal/types.h]: Ugh, fixed ;-(

File: include/kremlin/internal/types.h [hunk 4]
```diff
@@ -49,19 +49,21 @@ typedef const char *Prims_string;
 
 /* The great static header headache. */
 
+#define IS_MSVC64 (defined(_MSC_VER) && defined(_M_X64) && !defined(__clang__))
+#define HAS_INT128 ((defined(__x86_64__) || defined(__x86_64) || \
+                     defined(__aarch64__) || \
+                     (defined(__powerpc64__) && defined(__LITTLE_ENDIAN__)) || \
+                     defined(__s390x__)) || \
+                    (defined(_MSC_VER) && defined(__clang__)))
```
Comment (msprotz) [inline include/kremlin/internal/types.h]: could this last check be true in a hypothetical 32-bit build on windows with clang?

Comment (msprotz) [reply include/kremlin/internal/types.h]: which would then not have int128

File: kremlib/c/fstar_uint128_struct_endianness.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
+/* Verified uint128 implementation. */
```
Comment (msprotz) [inline kremlib/c/fstar_uint128_struct_endianness.h]: missing copyright header here

Comment (msprotz) [inline kremlib/c/fstar_uint128_struct_endianness.h]: also please update the comment: hand-written implementation of endianness-related uint128 functions for the extracted uint128 implementation

**Comment by msprotz** [comment]
this generally looks good -- our CI system has a bug so I'll have to push this to a branch on the fstarlang/kremlin repo, it looks like builds aren't succeeding for external forks

can you sign the CLA?

thanks!!

File: kremlib/c/fstar_uint128_struct_endianness.h (hunk 5 above)
Comment (msprotz) [reply kremlib/c/fstar_uint128_struct_endianness.h]: also missing header inclusion guard

File: include/kremlin/internal/types.h (hunk 4 above)
Comment (ueno) [reply include/kremlin/internal/types.h]: I've added a `!defined(_M_X64)` check here; would it be sufficient?

**Comment by ueno** [comment]
Thanks for the review; I've signed the CLA a while ago.

**Comment by msprotz** [comment]
thanks -- for some reason the CLA bot was complaining earlier

@beurdouche is there any chance you could give this a go at Mozilla to make sure it doesn't break any of your platforms? that way we can anticipate rather than break you next time you pull this code

**Comment by msprotz** [comment]
ping @beurdouche or @glandium would this fix the Firefox build on various platforms that Debian packages? I'm happy to merge this, I would just like a sanity check that this fixes the build issue

**Comment by glandium** [comment]
It fixes building NSS on Debian s390x. I haven't checked other architectures.

**Comment by msprotz** [comment]
Thanks Mike. I guess I'll just try to merge this and see what happens, unless you know how to run a candidate build for other Debian-supported platforms (I am no Debian developer so I have no clue).

**Comment by FlyGoat** [comment]
https://buildd.debian.org/status/package.php?p=nss&suite=sid
It fails on mips64el & riscv64 as well, should we add these two architectures to the if defined?

Thanks.

**Comment by msprotz** [comment]
I've pushed a working version of this PR based on @ueno's sketch (there were multiple issues, including a non-terminated #ifdef along with a #define defining some defined(...) which clang helpfully let me know is UB.

I also integrated @FlyGoat's suggestions for the other two platforms. This is on branch protz_feature_detection of kremlin.

@FlyGoat can you let me know if this fixes the build? I've also consolidated the ifdefs so that even if we think the platform doesn't support uint128, we always fallback to the verified (albeit, slow) implementation of uint128 using a pair of uint64. For some reason, this was not done before -- I'm hoping this makes the header more resilient and "always working".

Meanwhile, I'm running some Everest regression to make sure this doesn't break anything either, including in our compilation path with MSVC.

Thanks everyone for the feedback and suggestions, much appreciated.

**Comment by msprotz** [comment]
Everest CI came back green so I'm intending to merge protz_feature_detection at the end of the week.

**Comment by FlyGoat** [comment]
Can confirm it works on MIPS64EL.
Thanks!

**Comment by msprotz** [comment]
Wonderful, thanks for the help debugging!

File: include/kremlin/internal/types.h [hunk 6]
```diff
@@ -49,19 +49,21 @@ typedef const char *Prims_string;
 
 /* The great static header headache. */
 
+#define IS_MSVC64 (defined(_MSC_VER) && defined(_M_X64) && !defined(__clang__))
+#define HAS_INT128 ((defined(__x86_64__) || defined(__x86_64) || \
+                     defined(__aarch64__) || \
+                     (defined(__powerpc64__) && defined(__LITTLE_ENDIAN__)) || \
+                     defined(__s390x__)) || \
+                    (defined(_MSC_VER) && !defined(_M_X64) && defined(__clang__)))
```
Comment (glandium) [inline include/kremlin/internal/types.h:57]: Note that `__SIZEOF_INT128__` could be used to automatically add support to the platforms that aren't explicitly listed here but have compiler support for __int128.

Comment (msprotz) [reply include/kremlin/internal/types.h:57]: thanks, I've added that too in c3322295890570aeea0f240c0257910a907054de

**Comment by beurdouche** [comment]
This is now in NSS master (should be in 3.53). Thanks all !
https://github.com/nss-dev/nss/commit/184083346d77163ba2b8bd4ac8eb088618d2f36b

**Pull request merged**